tt_metal/impl/dispatch/arch.cpp

```diff
@@ -0,0 +1,738 @@
+        }
+        node_id_to_kernel.clear();
+    }
+
+    // Read the input table, create configs for each node
+    std::vector<dispatch_kernel_node_t> nodes = get_nodes(device_ids, num_hw_cqs);
+    for (const auto& node : nodes) {
+        TT_ASSERT(node_id_to_kernel.size() == node.id);
+        node_id_to_kernel.push_back(FDKernel::Generate(
+            node.id, node.device_id, node.cq_id, {node.my_noc, node.upstream_noc, node.downstream_noc}, node.kernel_type));
+    }
+
+    // Connect the graph with upstream/downstream kernels
+    for (const auto& node : nodes) {
+        for (int idx = 0; idx < DISPATCH_MAX_UPSTREAM; idx++) {
+            if (node.upstream_ids[idx] >= 0) {
+                // tt::log_info("Node {} has upstream node: {}", node.id, node.upstream_ids[idx]);
+                node_id_to_kernel.at(node.id)->AddUpstreamKernel(node_id_to_kernel.at(node.upstream_ids[idx]));
+            }
+        }
+        for (int idx = 0; idx < DISPATCH_MAX_DOWNSTREAM; idx++) {
+            if (node.downstream_ids[idx] >= 0) {
+                // tt::log_info("Node {} has downstream node: {}", node.id, node.downstream_ids[idx]);
+                node_id_to_kernel.at(node.id)->AddDownstreamKernel(node_id_to_kernel.at(node.downstream_ids[idx]));
+            }
+        }
+    }
+
+    // For kernels on mmio chip, need to confirm which remote device each is servicing
+    std::map<chip_id_t, uint32_t> device_id_to_tunnel_stop;
+    std::map<chip_id_t, std::vector<chip_id_t>> mmio_device_id_to_serviced_devices;
+    uint32_t tunnel_depth;
+    for (auto mmio_device_id : device_ids) {
+        if (tt::Cluster::instance().get_associated_mmio_device(mmio_device_id) != mmio_device_id)
+            continue;
+
+        // Get a list of remote devices serviced by this mmio chip
+        for (int idx = 0; idx < num_hw_cqs; idx++) {
+            mmio_device_id_to_serviced_devices[mmio_device_id].push_back(mmio_device_id);
+        }
+        std::vector<chip_id_t> remote_devices;
+        for (auto tunnel : tt::Cluster::instance().get_tunnels_from_mmio_device(mmio_device_id)) {
+            tunnel_depth = tunnel.size();
+            for (uint32_t tunnel_stop = 0; tunnel_stop < tunnel.size(); tunnel_stop++) {
+                chip_id_t remote_device_id = tunnel[tunnel_stop];
+                device_id_to_tunnel_stop[remote_device_id] = tunnel_stop;
+                if (remote_device_id != mmio_device_id) {
+                    for (int idx = 0; idx < num_hw_cqs; idx++) {
+                        remote_devices.push_back(remote_device_id);
+                    }
+                }
+            }
+        }
+
+        mmio_device_id_to_serviced_devices[mmio_device_id].insert(
+            mmio_device_id_to_serviced_devices[mmio_device_id].end(),
+            remote_devices.begin(),
+            remote_devices.end());
+    }
+
+    // Go through each mmio device, and assign the dispatch kernels on it to remote devices
```

Comment:
I don't understand this comment

Reply:
MMIO has some # of dispatch_h/prefetch_h kernels which service remote devices. To configure them we need to know the remote device the kernel is servicing, which is what the `SetServicingDeviceId()` calls populate.

Was wondering if it makes sense to have this as a field in the struct, although it would only apply to kernels on an mmio chip so not too sure

Reply:
I think that's ok.  We already have some "x" don't cares